Frameworks/CoreGraphics/DWriteWrapper.mm

```diff
@@ -580,29 +640,238 @@ HRESULT STDMETHODCALLTYPE CreateStreamFromKey(_In_ const void* fontFileReference
+        *hasCurrentFile = ++location < CFArrayGetCount(m_fontDatas.get());
+        return S_OK;
+    }
+
+private:
+    int location = -1;
+
+    woc::unique_cf<CFArrayRef> m_fontDatas;
+    std::vector<ComPtr<IDWriteFontFile>>* m_previouslyCreatedFiles;
+};
+
+/**
+ * Implementation of IDWriteFontCollectionLoader, which creates an IDWriteFontFileEnumerator to create font files
+ * Which are used to create a font collection.  This is expected to be a singleton
+ */
+class DWriteFontCollectionLoader : public RuntimeClass<RuntimeClassFlags<WinRtClassicComMix>, IDWriteFontCollectionLoader> {
+protected:
+    InspectableClass(L"Windows.Bridge.DirectWrite", TrustLevel::BaseTrust);
+
+public:
+    DWriteFontCollectionLoader() {
+    }
+
+    HRESULT RuntimeClassInitialize() {
+        m_fontDatas.reset(CFArrayCreateMutable(kCFAllocatorDefault, 0, &kCFTypeArrayCallBacks));
+        m_fontDatasSet.reset(CFSetCreateMutable(nullptr, 0, &kCFTypeSetCallBacks));
+        return S_OK;
+    }
+
+    void AddDatas(CFArrayRef fontDatas, CFArrayRef* errors) {
+        CFMutableArrayRef outErrors = nil;
+        if (errors) {
+            outErrors = CFArrayCreateMutable(nullptr, 0, &kCFTypeArrayCallBacks);
+            *errors = outErrors;
+        }
+
+        CFIndex count = CFArrayGetCount(fontDatas);
+        for (CFIndex i = 0; i < count; ++i) {
+            CFDataRef data = (CFDataRef)CFArrayGetValueAtIndex(fontDatas, i);
+            if (data) {
+                if (CFSetContainsValue(m_fontDatasSet.get(), data)) {
+                    if (errors) {
+                        woc::unique_cf<CFErrorRef> error{
+                            CFErrorCreate(nullptr, kCFErrorDomainCocoa, kCTFontManagerErrorAlreadyRegistered, nullptr)
+                        };
+                        CFArrayAppendValue(outErrors, error.get());
+                    }
+                } else {
+                    CFArrayAppendValue(m_fontDatas.get(), data);
+                    CFSetAddValue(m_fontDatasSet.get(), data);
+                    if (errors) {
+                        CFArrayAppendValue(outErrors, nullptr);
+                    }
+                }
+            } else {
+                if (errors) {
+                    woc::unique_cf<CFErrorRef> error{
+                        CFErrorCreate(nullptr, kCFErrorDomainCocoa, kCTFontManagerErrorInvalidFontData, nullptr)
+                    };
+                    CFArrayAppendValue(outErrors, error.get());
+                }
+            }
+        }
+    }
+
+    void RemoveDatas(CFArrayRef fontDatas, CFArrayRef* errors) {
+        CFMutableArrayRef outErrors = nil;
+        if (errors) {
+            outErrors = CFArrayCreateMutable(nullptr, 0, &kCFTypeArrayCallBacks);
+            *errors = outErrors;
+        }
+
+        CFIndex count = CFArrayGetCount(fontDatas);
+        for (CFIndex i = 0; i < count; ++i) {
+            CFDataRef data = (CFDataRef)CFArrayGetValueAtIndex(fontDatas, i);
+            if (data) {
```

Comment:
Both if can be joined. If(data && CFSetContains...)
To be fair, the nested ifs are growing in length throughout. It would be good to see if things can be factored or structured differently